# Example - running a SAP calculation with overwrite data

This notebook runs SAP10 calculation and overwrites some intermediate calculations with user supplied values.

This uses the [sap10calcs Python package](https://github.com/stevenkfirth/sap10calcs).

## 1. Setup

In [1]:
import sap10calcs

## 2. Run a normal SAP calculation

This shows that the dwelling is 'C' rated and the default gas fuel price used in SAP is 3.64 p/kWh.

In [2]:
response = sap10calcs.calculate(
    input_file = 'Example - creating a complete SAP XML file.xml',
    auth_token = None
)
output_dict = response.pop('sap_calculation_output_dict')
print('success:', response['sap_calculation_success'])
print('space_heating_main_system_1_other_fuel_price:', output_dict['space_heating_main_system_1_other_fuel_price'])
print('sap_value:', output_dict['sap_value'])
print('sap_band:', output_dict['sap_band'])


success: True
space_heating_main_system_1_other_fuel_price: 3.64
sap_value: 73.92082535021098
sap_band: C


## 3. Rerun the SAP calculation and overwrite the gas price

Let's look at gas prices from 1p/kWh to 5p/kWh

In [3]:
# this runs 5 SAP calculations
result = []
for i in range(1,6):
    response = sap10calcs.calculate(
        input_file = 'Example - creating a complete SAP XML file.xml',
        extra = {
            'overwrite': {
                'space_heating_main_system_1_other_fuel_price': i  # this overwrites the gas price
            }
        },
        auth_token = None
    )
    output_dict = response.pop('sap_calculation_output_dict')
    result.append(
        dict(
            space_heating_main_system_1_other_fuel_price = i,
            sap_value = output_dict['sap_value'],
            sap_band = output_dict['sap_band']
        
        )
    )
result

[{'space_heating_main_system_1_other_fuel_price': 1,
  'sap_value': 84.33375887248695,
  'sap_band': 'B'},
 {'space_heating_main_system_1_other_fuel_price': 2,
  'sap_value': 80.38946587162485,
  'sap_band': 'C'},
 {'space_heating_main_system_1_other_fuel_price': 3,
  'sap_value': 76.44517287076273,
  'sap_band': 'C'},
 {'space_heating_main_system_1_other_fuel_price': 4,
  'sap_value': 72.50087986990061,
  'sap_band': 'C'},
 {'space_heating_main_system_1_other_fuel_price': 5,
  'sap_value': 68.55658686903851,
  'sap_band': 'C'}]

This shows that varying the gas price has an impact on the SAP value and the SAP band.
- Setting the gas price to 1p/kWh increases the SAP value to 84.3 (a 'B' rating).
- Setting the gas price to 5p/kWh increases the SAP value to 68.6 (a 'C' rating).